distutils/ccompiler_opt.py: Werror is silently disabled

The compiler options are specified as three settings
https://github.com/numpy/numpy/blob/261a769dd135c1e4007a6d0bdbed68a8c929497a/numpy/distutils/ccompiler_opt.py#L193-L197

Those string values are then parsed
https://github.com/numpy/numpy/blob/261a769dd135c1e4007a6d0bdbed68a8c929497a/numpy/distutils/ccompiler_opt.py#L985-L998

with the sneaky self.cc_test_flags check that removes flags that compiler doesn't accept.

In our particular case Nix/Nixpkgs we have large user base that are concerned with security and thus a `-fno-strict-overflow` flag is being added in a compiler wrapper. This makes the check for `-Werror` flag support to wrongfully fail:
```
clang-7: error: argument unused during compilation: '-fno-strict-overflow' [-Werror,-Wunused-command-line-argument]
```

This results in `-Werror` flag being completely excluded in toolchain checks. In our recent update https://github.com/NixOS/nixpkgs/pull/132754 we've noticed that it fails to correctly detect availability of certain AVX512 intrinsics:

```
/private/tmp/nix-build-python3.9-numpy-1.21.1.drv-1/numpy-1.21.1/numpy/distutils/checks/extra_avx512bw_mask.c:12:11: error: implicit declaration of function '_kor_mask64' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    m64 = _kor_mask64(m64, m64);
          ^
```

### Reproducing code example:

TBD. Let me know if the description above is insufficient.

### Error message:

```
Original error was: dlopen(/nix/store/qhx0vgkjswsgp6gwn773mk0figasackk-python3.9-numpy-1.21.1/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-darwin.so, 2): Symbol not found: __cvtmask32_u32
  Referenced from: /nix/store/qhx0vgkjswsgp6gwn773mk0figasackk-python3.9-numpy-1.21.1/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-darwin.so
  Expected in: flat namespace
 in /nix/store/qhx0vgkjswsgp6gwn773mk0figasackk-python3.9-numpy-1.21.1/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-darwin.so
```

### NumPy/Python version information:

numpy 1.21.1

--- a/numpy/distutils/ccompiler_opt.py
+++ b/numpy/distutils/ccompiler_opt.py
@@ -193,7 +193,12 @@ class _Config:
         clang = dict(
             native = '-march=native',
             opt = "-O3",
-            werror = '-Werror'
+            # One of the following flags needs to be applicable for Clang to
+            # guarantee the sanity of the testing process, however in certain
+            # cases `-Werror` gets skipped during the availability test due to
+            # "unused arguments" warnings.
+            # see https://github.com/numpy/numpy/issues/19624
+            werror = '-Werror-implicit-function-declaration -Werror'
         ),
         icc = dict(
             native = '-xHost',

